src/util/PathUtil.cpp

```diff
@@ -1,63 +1,206 @@
+    gchar* filename = g_filename_from_uri(uri.c_str(), nullptr, nullptr);
+    if (filename == nullptr) {
+        return {};
     }
+    auto p = fs::u8path(filename);
+    g_free(filename);
 
-    g_error_free(error);
-    return false;
+    return p;
 }
 
-auto PathUtil::copy(const Path& src, const Path& dest) -> bool {
-    std::array<char, 16 * 1024> buffer{};  // 16k
+auto Util::toUri(const fs::path& path, GError** error) -> std::string {
+    char* uri = g_filename_to_uri(path.u8string().c_str(), nullptr, error);
 
-    FILE* fpRead = g_fopen(src.c_str(), "rbe");
-    if (!fpRead) {
-        return false;
+    if (uri == nullptr) {
+        return {};
     }
 
-    FILE* fpWrite = g_fopen(dest.c_str(), "wbe");
-    if (!fpWrite) {
-        fclose(fpRead);
-        return false;
+    string uriString(uri);
+    g_free(uri);
+    return uriString;
+}
+
+#ifndef BUILD_THUMBNAILER
+auto Util::fromGFile(GFile* file) -> fs::path {
+    char* p = g_file_get_path(file);
+    auto ret = p ? fs::u8path(p) : fs::path{};
+    g_free(p);
+    return ret;
+}
+
+auto Util::toGFile(const fs::path path) -> GFile* { return g_file_new_for_path(path.u8string().c_str()); }
+#endif
+
+
+void Util::openFileWithDefaultApplication(const fs::path& filename) {
+#ifdef __APPLE__
+    constexpr auto const OPEN_PATTERN = "open \"{1}\"";
+#elif _WIN32  // note the underscore: without it, it's not msdn official!
+    constexpr auto const OPEN_PATTERN = "start \"{1}\"";
+#else         // linux, unix, ...
+    constexpr auto const OPEN_PATTERN = "xdg-open \"{1}\"";
+#endif
+
+    string command = FS(FORMAT_STR(OPEN_PATTERN) % Util::getEscapedPath(filename));
+    if (system(command.c_str()) != 0) {
+        string msg = FS(_F("File couldn't be opened. You have to do it manually:\n"
+                           "URL: {1}") %
+                        filename.string());
+        XojMsgBox::showErrorToUser(nullptr, msg);
     }
+}
 
-    while (!feof(fpRead)) {
-        size_t bytes = fread(buffer.data(), 1, buffer.size(), fpRead);
-        if (bytes) {
-            fwrite(buffer.data(), 1, bytes, fpWrite);
-        }
+void Util::openFileWithFilebrowser(const fs::path& filename) {
+#ifdef __APPLE__
+    constexpr auto const OPEN_PATTERN = "open \"{1}\"";
+#elif _WIN32
+    constexpr auto const OPEN_PATTERN = "explorer.exe /n,/e,\"{1}\"";
+#else  // linux, unix, ...
+    constexpr auto const OPEN_PATTERN = R"(nautilus "file://{1}" || dolphin "file://{1}" || konqueror "file://{1}" &)";
+#endif
+    string command = FS(FORMAT_STR(OPEN_PATTERN) % Util::getEscapedPath(filename));
+    if (system(command.c_str()) != 0) {
+        string msg = FS(_F("File couldn't be opened. You have to do it manually:\n"
+                           "URL: {1}") %
+                        filename.string());
+        XojMsgBox::showErrorToUser(nullptr, msg);
     }
+}
+
+auto Util::getAutosaveFilepath() -> fs::path {
+    fs::path p(getConfigSubfolder("autosave"));
```

Comment:
```suggestion
    auto p = fs::u8path(getConfigSubfolder("autosave"));
```

Do this for the other getters too.

Reply:
getConfigSubfolder already returns a path, so it's redundant in this case. I've made sure to use fs::u8path when constructing the path from a glib function (e.g. g_get_user_config_dir()).

Reply:
exactly what I wanted :), just annotated the wrong line ^^.